### DISCUSSION TITLE
아이템, 동글몬, 차원 균열 알고리즘 설계

### DISCUSSION THREAD

**Discussion by jidoc01** [jidoc01/RustyDO#9] [Ideas]
현재 인게임에서 미구현된 중요 기능 3가지를 꼽자면, (1) 아이템, (2) 동글몬, 그리고 (3) 차원 균열입니다.
소스 코드 상에 관련 코드마다 코멘트는 달아놨지만, 어떤 알고리즘으로 출현/드롭시킬지 논의되지 않아서 미구현인 상태입니다.
이와 관련한 논의를 이 글에서 자유롭게 나누고자 합니다.

그리고, 물론 당시에 플레이했던 유저분들의 발언에 조금 더 무게가 실려야겠지만, 현실적으로 당시의 플레이어분들 찾는 것도 어렵지만 정확한 기억 가지신 분들 찾기도 어렵습니다 (세 분 계시면 세 분 기억이 다 다릅니다). 따라서 플레이 경험과 무관하게 자유롭게 의견 나눠주시면 좋을 것 같습니다. 자유롭게 의견을 나눈다는 건, 누구 말이 맞네 틀리네보다는 가급적이면 다양하게 의견을 서로 나눴으면 하는 바람입니다.

논의해야할 안건을 몇 가지 정해보자면...

+ 각각의 출현 주기는?
+ 각각의 출현 위치는?
+ 아이템의 경우, 종류에 따른 출현 빈도는?

이렇게 압축해볼 수 있을 것 같습니다.

## 출현 주기

얼마나 자주 출현할지 정해야 합니다. 예를 들어, 어떤 아이템이 반드시 한 사이클 (모든 플레이어가 한번씩 턴을 지나는 텀)마다 하나씩 나타날지,
아니면 두개씩 나타날지, 아니면 사이클 단위가 아닌 특정 수의 턴 (ex. 3턴?)마다 나타날지. 혹은 매 사이클마다 무작위한 순서의 턴에 출현할지.

위에서 말한 예시는 정말 가볍게 생각한 예시일 뿐입니다. 핵심은, 어떤 알고리즘으로 출현시킬지에 대한 구체적인 논의가 필요합니다. 그리고 그러한 알고리즘이 꼭 하나일 필요는 없습니다. 여러 아이디어를 종합해서 옵셔널하게 선택하게 할 수도 있으니까요.

## 출현 위치

어느 위치에 출현시킬지 정해야 합니다. 예를 들어, 맵의 영역을 몇개로 나누어서 영역 단위로 출현시킬지, 아니면 개별 위치 단위로 출현시킬지,
경기 운용에 아이템이 필요한 팀 근처에 좀 더 가중치를 높게 줘서 출현시킬지, 그랬을 때 경기 운용에 아이템이 필요한 팀에 대한 매트릭은 또 어떻게 짤지 등.

## 아이템 종류별 출현 빈도

아이템마다 출현하는 빈도가 달랐다고 들었습니다. 이와 관련해서 어떤 아이템에 높은 빈도를 부여할지 (흔한 아이템), 낮은 빈도를 부여할지 (희귀템)를 정할 수 있습니다. 이 부분은 아무래도 기억보다는 합리적으로 논의해서 결정하는 편이 좋을 것 같습니다.

그리고 아무래도 설계와 구현은 서로 동떨어질 수 없습니다만, 일단은 구현에 대한 논의보다는 하이레벨에서 설계를 우선적으로 자유롭게 논의하면 좋을 것 같습니다. 관련해서 구현을 아는 분들은 코멘트를 달아주시면 더 좋고요.

**Comment by jidoc01** [comment]
dhfdlsehdrjs님의 의견입니다:

+ 게임이 시작한지 5턴이 지나야 카드가 출현한다.
+ 5~7턴 간격으로 카드가 출현한다.
+ 카드는 특정 플레이어 앞에 떨어졌으나, 이동력 절반 정도를 소모해야 할 정도의 거리에 떨어졌다.
+ 상태 이상을 거는 카드들이 자주 나왔고, 그 다음이 진화 플러그인과 소형 회복 플러그인, 그 다음에 대형 플러그인이 나왔다. 암흑 진화 플러그인은 그 당시 확률이 엄청 낮았다.
+ 동글몬은 풀방 기준, 하나의 사이클이 돌면 나왔다. 유저가 적으면 2~3턴 정도만 지나도 나왔다.
+ 동글몬의 위치는 랜덤이었는데, 아이템과 마찬가지로 특정 유저 가까이에 출현했다.

스천알님의 의견입니다:

+ 차원 균열은 일정 턴이 지날 때마다 위치가 바뀌었다.
+ 5턴이 지나거나 상대 체력이 40프로 이하일 때, 동글몬 출현 아이템은 일정 확률로 카오스 진화 아이템이 나왔다 (의미 불분명).

**Comment by DigiEggz** [comment]
I think that there's enough information here to make a basic spawning algorithm. To condense the comments:

- Item is pulled from a pool based on % chance. These will need to be guessed since we have no way to determine original odds. The Dark Evolution item, for example, could be 5% (or less) chance.
-  Items spawn X pixels away from the player (X being about half of the movement bar). Did anyone weight in on if it felt like items were entirely random or instead given as support?
-  Culumon spawns every turn with many players (4+?), or every 2~3 with less players. Again, is the location random or based on who needs help?

For the point about the Chaos Evolution item, perhaps they meant that if the game has lasted 5+ turns and the user is injured, there's a small chance Culumon could evolve the players to a Dark evolution, functioning like the Darkness Training Manual item?

**Comment by mkyj100** [comment]
아직도 운영중이신가요?

**Reply by jidoc01** [reply]
서버 운영은 애초에 한 적이 없고, 개발은 제가 바쁘다 보니 현재 잠정 중단됐습니다

**Reply by mkyj100** [reply]
아하 답변 감사드립니다.
